### PR TITLE
Support Multiple Versions of Same Module at Runtime 

### PR DESCRIPTION
**Pull request by warunalakshitha**
## Purpose
This will provide the functionality of using multiple versions of same module in given ballerina code.

ex. 

```ballerina
import waruna/print version 1.0.0 as p1;
import waruna/print version 2.0.0 as p2;


public function main() {
    p1:printText("Hello World!");
    p2:printText("Hello World!");
 
}

````
Print module v1

```ballerina
import ballerina/io;

public function printText(string s) {
    io:println(s + " from v1");
}
```

Print module v2

```ballerina
import ballerina/io;

public function printText(string s) {
    io:println(s + " from v2");
}
```
output

```
Hello World! from v1
Hello World! from v2
```
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/21903

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [x] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by warunalakshitha** [comment]
ATM we can't test multiple versions of same module with balo dependencies since we can't add multiple versions of same import in ballerina.toml . Created https://github.com/ballerina-platform/ballerina-lang/issues/22984

File: gradle/balNativeLibProject.gradle [hunk 1]
```diff
@@ -149,7 +151,8 @@ task createBStringBalo(type: CreateBaloTask) {
     inputs.files(configurations.baloImplementation)
     inputs.files("$buildDir/build/classes/java/main/META-INF/this.map.json")
     outputs.dir("$buildDir/generated-balo")
-    outputs.files("$buildDir/generated-bir-jar/" + project.ext.orgName + "." + project.ext.moduleName + "-bstring.jar")
+    outputs.files("$buildDir/generated-bir-jar/" + project.ext.orgName + "." + project.ext.moduleName + "." + project
```
Comment (warunalakshitha) [inline gradle/balNativeLibProject.gradle:154]: Replce . with "-". for natives "java"

File: compiler/ballerina-lang/src/main/java/org/ballerinalang/codegen/BallerinaAnnotationProcessor.java [hunk 2]
```diff
@@ -282,8 +287,8 @@ protected String typeArrayToCode(List<TypeKind> types) {
         }
         
         public String code() {
-            return "registerNativeFunction(new NativeFunctionDef(\"" + this.org + "\", \"" + this.pkg + "\", " +
-                    "\"" + this.name + "\", " + this.typeArrayToCode(this.argTypes) + ", " +
+            return "registerNativeFunction(new NativeFunctionDef(\"" + this.org + "\", \"" + this.pkg + "\", \"" +
+                    this.version + "\", " + "\"" + this.name + "\", " + this.typeArrayToCode(this.argTypes) + ", " +
```
Comment (manuranga) [inline compiler/ballerina-lang/src/main/java/org/ballerinalang/codegen/BallerinaAnnotationProcessor.java:291]:
```suggestion
                    this.version + "\", \"" + this.name + "\", " + this.typeArrayToCode(this.argTypes) + ", " +
```

File: stdlib/grpc/src/main/java/org/ballerinalang/net/grpc/proto/ServiceProtoConstants.java [hunk 3]
```diff
@@ -47,6 +49,7 @@
     public static final String TMP_DIRECTORY_PATH = System.getProperty("java.io.tmpdir");
 
     public static final String ANN_SERVICE_CONFIG = "ServiceConfig";
+    public static final String ANN_SERVICE_CONFIG_FQN = PROTOCOL_STRUCT_PACKAGE_GRPC + ":" + "ServiceConfig";
```
Comment (manuranga) [inline stdlib/grpc/src/main/java/org/ballerinalang/net/grpc/proto/ServiceProtoConstants.java:52]:
```suggestion
    public static final String ANN_SERVICE_CONFIG_FQN = PROTOCOL_STRUCT_PACKAGE_GRPC + ":ServiceConfig";
```

File: gradle/balNativeLibProject.gradle (hunk 1 above)
Comment (warunalakshitha) [reply gradle/balNativeLibProject.gradle]: Fixed

File: cli/ballerina-tool/src/main/java/org/ballerinalang/tool/util/BFileUtil.java [hunk 4]
```diff
@@ -118,11 +118,18 @@ public FileVisitResult postVisitDirectory(Path dir, IOException exc) throws IOEx
...
      * @return Qualified class name
      */
-    public static String getQualifiedClassName(String orgName, String packageName, String className) {
+    public static String getQualifiedClassName(String orgName, String packageName, String version,
+                                               String className) {
+
+
         if (!Names.DEFAULT_PACKAGE.value.equals(packageName)) {
+            if (!"".equals(version)) {
+                className = version.replace('.', '_') + "." + className;
```
Comment (Kishanthan) [inline cli/ballerina-tool/src/main/java/org/ballerinalang/tool/util/BFileUtil.java:131]: looks like version.replace('.', '_') is a common code seen on multiple occurrences in this PR that can be moved to a method ?

File: langlib/lang.map/src/main/java/org/ballerinalang/langlib/map/ToArray.java [hunk 5]
```diff
@@ -43,7 +43,7 @@
  * @since 1.2.0
  */
 @BallerinaFunction(
-        orgName = "ballerina", packageName = "lang.map",
+        orgName = "ballerina", packageName = "lang.map", version = "1.1.0",
```
Comment (Kishanthan) [inline langlib/lang.map/src/main/java/org/ballerinalang/langlib/map/ToArray.java:46]: for lang.lib module versions, can we use constants and refer them wherever needed?

File: tests/jballerina-integration-test/src/test/java/org/ballerinalang/test/packaging/MultipleVersionsModuleTestCase.java [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+import static org.ballerinalang.test.packaging.PackerinaTestUtils.deleteFiles;
+import static org.wso2.ballerinalang.compiler.util.ProjectDirConstants.BLANG_COMPILED_JAR_EXT;
+import static org.wso2.ballerinalang.compiler.util.ProjectDirConstants.BLANG_COMPILED_PKG_BINARY_EXT;
+
+/**
+ * Testing usage of multiple versions of same module.
+ *
+ * @since 2.0.0
+ */
+public class MultipleVersionsModuleTestCase extends BaseTest {
```
Comment (warunalakshitha) [inline tests/jballerina-integration-test/src/test/java/org/ballerinalang/test/packaging/MultipleVersionsModuleTestCase.java:51]: @Kishanthan Testcase for multiple versions

File: cli/ballerina-tool/src/main/java/org/ballerinalang/tool/util/BFileUtil.java (hunk 4 above)
Comment (warunalakshitha) [reply cli/ballerina-tool/src/main/java/org/ballerinalang/tool/util/BFileUtil.java:131]: even it is multiple occurred, these are in different modules and packages. Then moving to method will not have any advantage and this is also inside util method of each module

**Pull request merged**